### DISCUSSION TITLE
Blending AnimStarterPack montages

### DISCUSSION THREAD

**Discussion by Jventoo** [Sixze/ALS-Refactored#19] [Q&A] [answered]
Hi,

AI engineer without much animation experience here so please bear with me 😃

I was looking at the layering blending that goes on in the animation blueprint. There's some really helpful comments in the ABP about blending upper body montages like reloads and I was wondering what the proper way to set the curves slot up was so that montages, like from Epic's AnimStarterPack for example, will properly blend with the various overlay states. The montage works fine for the default overlay state, but not so much with ones that modify the arms slots like M4.

Setup for my reload montage is as follows:

![image](https://user-images.githubusercontent.com/3110882/133707760-a35d5609-b1f4-499a-b51e-c446507aa105.png)

Let me know if a video of the strange layer blending I was describing would be helpful! Thanks!

**Comment by Sixze** [comment, marked as answer]
This is how layering basically works:

![1](https://user-images.githubusercontent.com/16257871/133818875-a187de1a-6979-4b26-a037-c282a4636b2d.png)

1. Blend the overlay pose and the montage pose using the "Slots" layering curves.
2. Additive apply the locomotion pose on the pose from the previous step using the "Additive" layering curves.
3. Blend the pose from the previous step and the locomotion pose using regular layering curves.

To use blending effectively, you need to have all layering curves for all body parts in your animation, even if most of the curves remain at 0.

There is a very useful animation modifier called `UAlsAnimationModifier_CreateLayeringCurves` that automatically creates all the curves you need, but animation montages cannot have modifiers, so you must use it on a regular animation sequence.

So, this is how I do things like this:

1. In addition to the regular animation, I create two additive versions of it.

![image](https://user-images.githubusercontent.com/16257871/133819054-ed9e3bdc-fd49-46a8-9a1a-6d661f2c5314.png)

2. Apply `UAlsAnimationModifier_CreateLayeringCurves` to regular animation only, no curves are needed for additive animation.

![image](https://user-images.githubusercontent.com/16257871/133819165-a4676024-e69a-4e49-ada2-a95a072ed56f.png)

3. Assign animations to layering slots. You can experiment with this, but always assign animation with curves to the "Layering.Curves" slot.

![image](https://user-images.githubusercontent.com/16257871/133819282-0e7fe2ac-e874-4b7a-a98f-ea7b4adf979f.png)

4. Experiment with curves :).

If you still have questions, then, as an example, I can give you an animation from a project that I am working on.

**Reply by craigrushforth** [reply]
This is great tip 👍🏻   Going to give this a go later tonight.  I really like what you've done with ALS-Refactored.  I've already learnt so much just looking through the code.  Love that you have the GameplayTag to switch overlays and there is even some ControlRig going on at the end.  

If anyone has any tips / documentation / video links please send me the details.  I'm new to ALS and as there are so many different versions its hard tracking down the best practices.

A lot of the stuff that looks incredible seems to be paid monolithic solutions but I'd be happy with a handful of small concise examples.

E.g. 
* An upper body montage (Looks like the example above is a perfect description of that)
* If I wanted the character to point a flash light is that an overlay that also has poses for the look / aim directions
* If I wanted a zombie do you do a full body overlay to get the twitchy zombieness but also an overlay for the attack animations (probably not since you can only have a single overlay.  So I suppose that would fall under a Gait or something).

I'm still new and trying to figure out where the limits of Actions / Gaits / Overlays begin and end.

Also I'm using the Gameplay Ability System so if anyone has any tips on how to combine the two that'd be really useful.

**Reply by alexanderzinolidis** [reply]
I am also having some difficulty in understanding how to add firing animations. I'm experimenting but would love to see some examples of this with something like the AnimStarterPack.

**Reply by assmikDuDuDuJ** [reply]
@Sixze Can I have an example?

**Reply by VioLence1982** [reply]
Can u make a Tutorial video im to Stupid to get my Animations to work
 :(

**Reply by EddieSoloProjects** [reply]
Would love to see a video Tut on this

**Reply by Ender0920** [reply]
@Sixze Hi，can I have an example animation,please~

**Reply by Deltalo** [reply]
Is this working in 5.6? I just downloaded and tested the method described here but I’m not getting any results. I’m leaving my configuration here:

<img width="419" height="179" alt="image" src="https://github.com/user-attachments/assets/81bbe51a-0bac-429f-8784-faf0de307e79" />

I also added the curves with the modifier here (only in the normal animation , not in the additive version)

<img width="1436" height="1169" alt="image" src="https://github.com/user-attachments/assets/e3c06217-c035-42c3-ad6c-ca3db27d5e0f" />

And my montage ended up like this:

<img width="1421" height="1097" alt="image" src="https://github.com/user-attachments/assets/046f7465-777d-4797-8cb3-1c25749abcbd" />

But I have this final result, Here just a montage play to test :



https://github.com/user-attachments/assets/9d5a22ae-c8ed-4ba1-86d7-274054c08778

**Reply by ameaninglessname** [reply]
You could analyze it further with rewind debugger.

**Reply by Deltalo** [reply]
Ohh thank you so much, with that I discovered that I was missing the layering slots, in the curves , Now with this it works perfectly, I just need to adjust it so it looks good @ameaninglessname!!

**Comment by T6nuOss** [comment]
@Sixze Hi, I would like to have an example animation aswell.

**Reply by T6nuOss** [reply]
Nvm. After 3 evenings failure I asked example and after 5 min asking example I figured it out.

**Reply by pbaughman88** [reply]
How?
